src/Splat.Microsoft.Extensions.DependencyInjection/MicrosoftDependencyResolver.cs

```diff
@@ -0,0 +1,213 @@
+        public void UnregisterAll(Type serviceType, string contract = null)
+        {
+            /*
+            if (_isImmutable)
+            {
+                // replaced so to provide a way to remove scopes
+                // ugly but best I could do
+                throw new InvalidOperationException(ImmutableExceptionMessage);
+            }
+            */
+
+            if (_isImmutable)
+            {
+                var scopes = _serviceScopes.Value;
+
+                if (contract == null || !scopes.TryGetValue(contract, out var scope))
+                {
+                    throw new ArgumentOutOfRangeException($"The contract {contract} was not instantiated.");
+                }
+
+                scope.Dispose();
+                scopes.Remove(contract);
+                return;
+            }
+
+            if (serviceType == null)
+            {
+                throw new ArgumentNullException(nameof(serviceType));
+            }
+
+            var registrations = _serviceCollection
+                .Where(sd => sd.ServiceType == serviceType)
+                .ToList();
+
+            if (!registrations.Any())
+            {
+                throw new ArgumentOutOfRangeException($"The type {serviceType} was not registered.");
+            }
+
+            foreach (var sd in registrations)
+            {
+                _serviceCollection.Remove(sd);
+            }
+        }
+
+        /// <inheritdoc />
+        public IDisposable ServiceRegistrationCallback(Type serviceType, string contract, Action<IDisposable> callback)
+        {
+            // this method is not used by RxUI
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc/>
+        public bool HasRegistration(Type serviceType)
+        {
+            if (!_isImmutable)
+            {
+                return _serviceCollection.Any(sd => sd.ServiceType == serviceType);
+            }
+
+            var service = _serviceProvider.GetService(serviceType);
+            return service != null;
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Disposes of the instance.
+        /// </summary>
+        /// <param name="disposing">Whether or not the instance is disposing.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+        }
+
+        private IServiceProvider GetServiceProvider(string contract = null)
+        {
+            if (_serviceProvider == null)
+            {
+                _serviceProvider = _serviceCollection.BuildServiceProvider();
+            }
+
+            var dic = _serviceScopes.Value;
+
+            if (contract == null)
+            {
+                return _serviceProvider;
+            }
+
+            if (!dic.TryGetValue(contract, out var scope))
```

Comment:
This dictionary doesn't seem to be populated earlier?

Reply:
Do you want me to change it so that the dictionary holds types per contract and uses scope for resolving them?
As I commented [above](https://github.com/reactiveui/splat/pull/370/files/f5c2900d99af904e5ff78b4ba6be47eaf6f9d5ae#r309722126), MS DI doesn't offer to register keyed items or scopes. If we want to achieve that we have to maintain that on our own, do we want to?

Reply:
@weitzhandler This is the issue I ran into when I implemented Simple Injector and what I was working out on my MSDI branch.  I was working on a dictionary to allow for unregistering and named instances.  I was going to use MSDI as the test that it would work then implement it for Simple Injector.

Reply:
@RLittlesII 

> @weitzhandler This is the issue I ran into when I implemented Simple Injector

Yes, I can see `SimpleInjector` complete ignores `contract` as well.

It's going to be an issue with MS DI, which does not allow registrations to be keyed by string, only by types. I'm not sure how to achieve this, because scope is only a clone of the original registration chart, not a mutation of it. If we could only generate a dynamic type from a string, but I don't think that's an option, definitely not in .NET Standard.
Any ideas, or do I completely remove the scopes thing for the moment?

Reply:
See how to do it below. Don't care what SimpleInjector does, we need to open a issue for it to be fixed there also. This needs to be fixed.

Reply:
Exactly.  Simple Injector has an open issue to fix it.  The solution should be similar for both.  We need to keep an internal dictionary that has those types and keep the underlying service registry in sync with it.  I believe there are other ways to do it.  That was the road I went down.